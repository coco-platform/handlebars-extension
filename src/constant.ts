/**
 * @description - declare specific types
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

/* tslint:disable:max-classes-per-file */

export class Medium {
  public result: string[];
}

export enum ResourceType {
  stylesheet = 'stylesheet',
  script = 'script',
}

export class InlineResource {
  public type: ResourceType;
  public content: string;
}

export class ScriptNode {
  public type: string;
  public src?: string;
  public async?: string;
  public defer?: string;
  public crossorigin?: string;
}

export class LinkNode {
  public type: string;
  public href?: string;
  public rel?: string;
}

export function isString(value: any) {
  return typeof value === 'string';
}
