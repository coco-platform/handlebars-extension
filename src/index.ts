/**
 * @description - nothing but export bundle
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

export { inline } from './inline';
export { link } from './link';
