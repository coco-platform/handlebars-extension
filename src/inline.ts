/**
 * @description - handlebars inline expression
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

/**
 * @typedef {object} InlineResource
 *
 * @property {string} type - stylesheet, script only
 * @property {string} content
 */

// package
import { SafeString, Utils } from 'handlebars';
import { InlineResource, Medium, ResourceType } from './constant';
// scope
const { escapeExpression } = Utils;

/**
 * @param {InlineResource} resource
 *
 * @return {SafeString|string}
 */
export function inline(resource?: InlineResource): SafeString | string {
  const state: Medium = { result: [] };
  const { type, content } = resource;

  switch (type) {
    case ResourceType.stylesheet:
      state.result = [
        '<style type="text/css">',
        escapeExpression(content),
        '</style>',
      ];
      break;
    case ResourceType.script:
      state.result = [
        '<script type="application/javascript">',
        escapeExpression(content),
        '</script>',
      ];
      break;
  }

  return state.result.length ? new SafeString(state.result.join('\n')) : '';
}
