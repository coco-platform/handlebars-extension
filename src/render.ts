/**
 * @description - render AST into string
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

// package
import { SafeString } from 'handlebars';
import { isString, ResourceType, ScriptNode } from './constant';

export function render(type: ResourceType, resource: ScriptNode): SafeString {
  const tag =
    type === ResourceType.script
      ? '<script type="application/javascript"></script>'
      : '<link type="text/css">';
  const props = Reflect.ownKeys(resource)
    .map((key) => [key, Reflect.get(resource, key)])
    .reduce(
      (acc, [key, value]) =>
        isString(value) ? `${acc} ${key}="${value}"` : `${acc} ${key}`,
      ''
    );
  const node = tag.replace('>', `${props}>`);

  return new SafeString(node);
}
