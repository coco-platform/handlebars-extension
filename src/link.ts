/**
 * @description - handlebars link expression
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

/**
 * @typedef {object} LinkNode
 *
 * @property {string} [href]
 * @property {string} [rel]
 */

/**
 * @typedef {object} ScriptNode
 *
 * @property {string} [src]
 * @property {boolean} [async]
 * @property {boolean} [defer]
 * @property {boolean} [crossorigin]
 */

// package
import { SafeString } from 'handlebars';
import { LinkNode, ResourceType, ScriptNode } from './constant';
import { render } from './render';

/**
 * @param {LinkNode|ScriptNode} resource
 *
 * @return {SafeString|string}
 */
export function link(resource?: ScriptNode | LinkNode): SafeString | string {
  const scriptRegExp = /\.js$/;
  const stylesheetRegExp = /\.css$/;

  switch (true) {
    case scriptRegExp.test((resource as ScriptNode).src):
      return render(ResourceType.script, resource);
    case stylesheetRegExp.test((resource as LinkNode).href):
    case Boolean((resource as LinkNode).rel): // prefetch, preload
      return render(ResourceType.stylesheet, resource);
    default:
      return '';
  }
}
