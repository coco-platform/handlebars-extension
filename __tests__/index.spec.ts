/**
 * @description - handlebars extensions
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

// package
import { readFileSync } from 'fs';
import * as hbs from 'handlebars';
import { resolve } from 'path';
// internal
import { inline, link } from '../src';
// scope
const options = { encoding: 'utf8' };

describe('@coco-platform/handlebars-extension', () => {
  beforeAll(() => {
    hbs.registerHelper('inline', inline);
    hbs.registerHelper('link', link);
  });

  it('should support inline expression', () => {
    const demo = resolve(__dirname, '__fixture__', 'inline.hbs');
    const template = readFileSync(demo, options);
    const context = {
      criticals: [
        {
          type: 'stylesheet',
          content:
            '.CloseIcon-icon-2xww{transition:opacity .3s ease-out}.CloseIcon-icon-2xww:hover{opacity:.8}',
        },
        {
          type: 'script',
          content: `(function($) {$(document).ready(() => {});})(jQuery);`,
        },
        {},
      ],
    };
    const result = hbs.compile(template)(context);

    expect(result).toMatchSnapshot();
  });

  it('should support link expression', () => {
    const demo = resolve(__dirname, '__fixture__', 'link.hbs');
    const template = readFileSync(demo, options);
    const context = {
      resources: [
        {
          href: 'https://static.zhihu.com/heifetz/main.css',
        },
        {
          rel: 'dns-prefetch',
          href: `//static.zhimg.com`,
          async: true,
        },
        {
          src: 'https://static.zhihu.com/heifetz/main.js',
          defer: true,
          crossorigin: 'anoymous',
        },
        {},
      ],
    };
    const result = hbs.compile(template)(context);

    expect(result).toMatchSnapshot();
  });
});
