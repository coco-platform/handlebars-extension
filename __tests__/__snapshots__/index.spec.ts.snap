// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@coco-platform/handlebars-extension should support inline expression 1`] = `
"  <style type=\\"text/css\\">
.CloseIcon-icon-2xww{transition:opacity .3s ease-out}.CloseIcon-icon-2xww:hover{opacity:.8}
</style>
  <script type=\\"application/javascript\\">
(function($) {$(document).ready(() &#x3D;&gt; {});})(jQuery);
</script>
  
"
`;

exports[`@coco-platform/handlebars-extension should support link expression 1`] = `
"  <link type=\\"text/css\\" href=\\"https://static.zhihu.com/heifetz/main.css\\">
  <link type=\\"text/css\\" rel=\\"dns-prefetch\\" href=\\"//static.zhimg.com\\" async>
  <script type=\\"application/javascript\\" src=\\"https://static.zhihu.com/heifetz/main.js\\" defer crossorigin=\\"anoymous\\"></script>
  
"
`;
